components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Runtime/Extensions/CertificateStoreEx.cs

```diff
@@ -4,45 +4,83 @@
+                throw new ArgumentNullException(nameof(certificateTrustList));
+            }
+
+            if (certificateTrustList.StorePath != certificateStore.StorePath) {
+                certificateTrustList.StoreType = certificateStore.StoreType;
+                certificateTrustList.StorePath = certificateStore.StorePath;
+            }
 
             return certificateTrustList;
         }
 
         /// <summary>
-        /// Create identifier
+        /// Build the certificate identifier for the own app
         /// </summary>
-        /// <param name="certificateInfo"></param>
-        /// <param name="hostname"></param>
+        /// <param name="certificateStore"></param>
+        /// <param name="certificateIdentifier"></param>
         /// <returns></returns>
-        public static CertificateIdentifier ToCertificateIdentifier(
-            this CertificateInfo certificateInfo, string hostname) {
-            var certificateIdentifier = new CertificateIdentifier {
-                StoreType = certificateInfo.StoreType,
-                StorePath = certificateInfo.StorePath,
-                SubjectName = certificateInfo.SubjectName.Replace("DC=localhost", $"DC={hostname}")
-            };
+        public static CertificateIdentifier BuildCertificateIdentifier(
+            this CertificateInfo certificateStore,
+            CertificateIdentifier certificateIdentifier) {
+            if (certificateStore == null) {
+                throw new ArgumentNullException(nameof(certificateStore));
+            }
+
+            if (certificateIdentifier == null) {
+                throw new ArgumentNullException(nameof(certificateIdentifier));
+            }
+
+            if (certificateIdentifier.StorePath != certificateStore.StorePath) {
+                certificateIdentifier.StoreType = certificateStore.StoreType;
+                certificateIdentifier.StorePath = certificateStore.StorePath;
+            }
+
             return certificateIdentifier;
         }
+
+        /// <summary>
+        /// Build a certificate identifier store. Particularily for rejected certificates
+        /// </summary>
+        public static CertificateStoreIdentifier BuildCertificateIdentifierStore(
+            this CertificateStore certificateStore,
+            CertificateStoreIdentifier certificateStoreIdentifier) {
+            if (certificateStore == null) {
+                throw new ArgumentNullException(nameof(certificateStore));
+            }
+
+            if (certificateStoreIdentifier == null) {
+                throw new ArgumentNullException(nameof(certificateStoreIdentifier));
+            }
+
+            if (certificateStoreIdentifier.StorePath != certificateStore.StorePath) {
+                certificateStoreIdentifier.StoreType = certificateStore.StoreType;
+                certificateStoreIdentifier.StorePath = certificateStore.StorePath;
+            }
+
+            return certificateStoreIdentifier;
```

Comment:
This is just a helper because the publisher allows to specifiy all cert stores, now just the pkiroot is given and the rest autopopulated. 367 also had a bug that it didn't pick up the root for the rejected store...

Reply:
Even for a helper method it is confusing that input parameter is being changed and return values is not used.

Reply:
I will refactor this soon.